### DISCUSSION TITLE
Provenance should be more prominent

### DISCUSSION THREAD

**Discussion by Jack-Works** [community/community#128399] [npm]
Currently, the provenance is listed under the readme, IMO it should be in the sidebar.

**Comment by feelepxyz** [comment]
@Jack-Works appreciate the feedback! We're thinking of ways to improve this.

**Reply by JoshBashed** [reply]
I think you should move it to the top of the page. Additionally, if the package does not have provenance, there should be a box like this.
![image](https://github.com/npm/feedback/assets/36625023/16adfbd1-cf95-4738-92df-7b3691dd105b)

**Reply by ljharb** [reply]
Provenance doesn't verify "source" whatsoever, though - it only verifies the publish location, and nothing more.

**Reply by JoshBashed** [reply]
@ljharb yea, it verifies the the repo.

**Reply by ljharb** [reply]
Correct - but the contents of the repo are in no way guaranteed to be connected to the contents of the published package.

**Reply by JoshBashed** [reply]
> Correct - but the contents of the repo are in no way guaranteed to be connected to the contents of the published package.

This feature actually mentions the tag/commit id of the repo used in the build process!

**Reply by ljharb** [reply]
Sure - but the contents often won't match that, since they're often the output of a build process - and they could be downloaded from an arbitrary URL as well. That it will *happen* to be mostly true for most cases doesn't mean provenance is providing you any guarantee about it whatsoever.

**Reply by JoshBashed** [reply]
> Sure - but the contents often won't match that, since they're often the output of a build process - and they could be downloaded from an arbitrary URL as well. That it will *happen* to be mostly true for most cases doesn't mean provenance is providing you any guarantee about it whatsoever.

The point of this feature is to verify that the contents of an npm package is the result of the build pipeline. The build pipeline and the code at the time of the build is both saved. Because of that, anyone can verify that the code is the result of the build pipeline without any *unexpected* code added. If a build pipeline downloads arbitrary files from arbitrary URLs it must be part of the build steps that are saved.

**Reply by ljharb** [reply]
Sure, that's totally true. That still means that provenance doesn't guarantee anything but the repo/workflow file from which the publish occurred. The rest *may*, or may not, be determinable from that workflow - but that's separate from provenance itself.

**Reply by JoshBashed** [reply]
> Sure, that's totally true. That still means that provenance doesn't guarantee anything but the repo/workflow file from which the publish occurred. The rest may, or may not, be determinable from that workflow - but that's separate from provenance itself.

I am sure you *could* create a system that would download additional code and inject it into a library, but, this would all be shown in the build steps.

**Reply by ljharb** [reply]
and, since you *could*, that means provenance doesn't guarantee it. That a human could manually go determine that isn't any different than without provenance.

**Reply by JoshBashed** [reply]
> and, since you could, that means provenance doesn't guarantee it. That a human could manually go determine that isn't any different than without provenance.

I am not sure why you trying to argue off technicalities. The point of this feature is to have a zero trust way to verify the build process, the repo that it's built from, and the steps used to build it. You *could* add those build steps that are malicious, but, everyone could see that publicly by clicking one link [(here is a highlight)](https://www.npmjs.com/package/rfil#:~:text=.github/workflows/publish.yaml).

**Reply by ljharb** [reply]
The details are critically important. You can programmatically verify provenance, but that won't tell you anything about the build unless a _human_ looks at it.

**Reply by MylesBorins** [reply]
I think the critical difference is that provenance gives you the bread crumb trail that you know is verified. Without provenance you have nothing.

As per oxford dictionary

provenance is "the place of origin or earliest known history of something."

The feature doesn't claim to be anything else.

**Reply by ljharb** [reply]
the npm website's badge says "build provenance" which is the problematic claim, "provenance" or "publish provenance" i agree are accurate terms.

**Reply by JoshBashed** [reply]
So if you don't think provenance verifies the source, what does it verify?

**Reply by ljharb** [reply]
@JoshuaBrest the only thing it verifies is the publish location (which, for github, includes a repo and a github action workflow file/sha)

**Reply by feelepxyz** [reply]
> the npm website's badge says "build provenance" which is the problematic claim

This is a fair point and maybe we should say something like `Published and signed on` 🤔 

> @JoshuaBrest the only thing it verifies is the publish location (which, for github, includes a repo and a github action workflow file/sha)

It might be helpful to dig into the details a bit here. npm provenance is leveraging claims from the [OIDC ID token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect) that CI/CD providers like Actions and GitLab expose as part of a workflow/pipeline job. This contains signed metadata about the triggering event. For example, you have a workflow that publishes when you push to a branch. The provenance will include this metadata about the triggering event that kicked off the job that published the package. 

If you have a basic workflow that just runs `actions/checkout` and `npm publish` you will be building and publishing from the exact source commit that's linked. But as always, there are many many ways of doing something funky, like checking out a different commit/branch, or run a build script that downloads stuff and injects it in the package.

Provenance does not make any guarantees about what happens during build, not does it protect against source code compromise which is often a pre-requisite for these types of attacks. What it gives you is a verifiable pointer to the originating trigger/workflow that published the package. This should be seen as an "identity", similar to who the maintainer is. 

This gives you another signal to trust vs just that a authorized maintainer published the package. An attacker would now need to take over both your npm account and github account! This is only getting harder with maintainers adopting 2FA, turning on branch protections etc. 

Take [`create-next-app`](https://www.npmjs.com/package/create-next-app) for example. If you know that `vercel` maintain it you would expect the package to come from a repo like `vercel/create-next-app` (in this case it's a monorepo `vercel/next.js`). If suddenly it comes from a different user/org, there's cause for concern, and with provenance we have a way of detecting these types of changes.

**Reply by feelepxyz** [reply]
> Additionally, if the package does not have provenance, there should be a box like this.

I'm not sure we want to go down this road for packages that don't have provenance as this would penalise legacy packages that don't need/require any updates (something @ljharb pointed out previously). Take [lodash](https://www.npmjs.com/package/lodash) as an example, it's pretty stable and was last released 2 years ago. 

What I'm envisioning are more ways of working with provenance as a trusted identity for a package, e.g. warn when it changes when updating dependencies, allow users to have policies that enforce it for certain packages, or block unexpected changes during publish/force additional verification to change.

**Reply by ljharb** [reply]
Since you're publishing from github with a stored npm secret, in fact it means they'd only need to take over your github account (which lets them bypass branch protections in most cases), not your npm account.

**Reply by JoshBashed** [reply]
> What I'm envisioning are more ways of working with provenance as a trusted identity for a package, e.g. warn when it changes when updating dependencies, allow users to have policies that enforce it for certain packages, or block unexpected changes during publish/force additional verification to change.

Maybe only show for for newly updated/created packages. @feelepxyz

**Comment by panva** [comment]
I've learned (from a private discussion with @MylesBorins) that publishing with provenance cross checks the URLs (at the very least repository?) in package.json.

This means that the same check mark can be placed next to the repository sidebar section as well.

Edit: I would definitely appreciate a check list of cross checks the publishing does listed somewhere because this point wasn't apparent from the announcement or the docs on npmjs.

**Reply by feelepxyz** [reply]
@panva yeah we're thinking of ways to make this information easier to find and tying this to the repo could make sense.

> I would definitely appreciate a check list of cross checks the publishing does listed somewhere

👍  we might codify this in a spec eventually, but for now, the registry does the following checks before accepting a publish with provenance:
- Verifies the provenance attestation is for the particular package, name and tarball shasum being published (this info is in the signed attestation subject) (open the `Attestation` drawer to see the `subject` on [this page](https://search.sigstore.dev/?logIndex=18817304))
- Verifies the provenance details (repo, ref, sha etc) match what's in the signing certificate, and what was signed by the CI/CD system (check the values in the [Public Key Certificate here](https://search.sigstore.dev/?logIndex=18817304))
- Verifies the repository in `package.json` is set and matches the one in the signed provenance attestation
- Verifies the signed provenance bundle using sigstore-js verify: https://github.com/sigstore/sigstore-js/#verifybundle-payload-options (checks it was signed by Sigstore, was signed during publish etc)
- Verifies the CI/CD system is supported (currently GitHub Actions, soon [GitLab](https://github.com/npm/cli/pull/6375) and hopefully others)

**Reply by feelepxyz** [reply]
When you run `npm audit signatures` on installed packages with provenance we run the following checks:
- Verifies the provenance attestation is for the particular package, name and tarball shasum that was installed (this info is in the signed attestation subject) (open the Attestation drawer to see the subject on [this page](https://search.sigstore.dev/?logIndex=18817304))
- Verifies the signed provenance bundle using sigstore-js verify: https://github.com/sigstore/sigstore-js/#verifybundle-payload-options (checks it was signed by Sigstore, was signed during publish etc)
- Verifies the npm registry accepted the publish with provenance (indirectly asserting all the above checks)

**Reply by carmithershman** [reply]
Hello @feelepxyz, can you please point me to a documentation of how to add integration to npm --provenance as CI/CD and registries vendor? I would like to look into this

**Reply by feelepxyz** [reply]
@carmithershman yes I'm planning to write something up integrating provenance into a registry vendor, I'll let you know when we have an early draft. We've also got a work in progress document for CI/CD vendors that I'm hoping we can share soon, maybe next week.

**Comment by carmithershman** [comment]
Hi Philip

Thank you, I am waiting for your update

Best Regards
Carmit Hershman
